core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadHandler.java

```diff
@@ -312,25 +273,189 @@ public CompleteMultipartUploadResult call() throws S3Exception {
+              status -> Integer.parseInt(status.getName()),
+              status -> status
+      ));
+      int lastPartNum = request.getParts().get(request.getParts().size() - 1).getPartNumber();
+      for (CompleteMultipartUploadRequest.Part part : request.getParts()) {
+        if (!uploadedPartsMap.containsKey(part.getPartNumber())) {
+          throw new S3Exception(objectPath, S3ErrorCode.INVALID_PART);
+        }
+        if (part.getPartNumber() != lastPartNum // size requirement not applicable to last part
+                && uploadedPartsMap.get(part.getPartNumber()).getLength() < Configuration.getBytes(
+                PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_MIN_PART_SIZE)) {
+          throw new S3Exception(objectPath, S3ErrorCode.ENTITY_TOO_SMALL);
+        }
+      }
+      return uploadedParts;
+    }
+
+    /**
+     * Cleanup the multipart upload temporary folder holding the parts files.
+     * and the meta file for this multipart.
+     * @param multipartTemporaryDir
+     * @throws IOException
+     * @throws AlluxioException
+     */
+    public void removePartsDirAndMPMetaFile(AlluxioURI multipartTemporaryDir)
+            throws IOException, AlluxioException {
+      mUserFs.delete(multipartTemporaryDir,
+              DeletePOptions.newBuilder().setRecursive(true).build());
+      mMetaFs.delete(new AlluxioURI(
+                      S3RestUtils.getMultipartMetaFilepathForUploadId(mUploadId)),
+              DeletePOptions.newBuilder().build());
+      if (mMultipartCleanerEnabled) {
+        MultipartUploadCleaner.cancelAbort(mMetaFs, mUserFs, mBucket, mObject, mUploadId);
+      }
+    }
+
+    /**
+     * Cleanup the temp object file for complete multipart upload.
+     * @param objTempPath
+     */
+    public void cleanupTempPath(String objTempPath) {
+      if (objTempPath != null) {
+        try {
+          mUserFs.delete(new AlluxioURI(objTempPath), DeletePOptions.newBuilder().build());
+        } catch (Exception e) {
+          LOG.warn("Failed to clean up temp path:{}, {}", objTempPath, e.getMessage());
+        }
+      }
+    }
+
+    /**
+     * On any exception, check with Master on if the there's a object file.
+     * bearing the same upload id already got completed.
+     * @param objectPath
+     * @return the status of the existing object through CompleteMultipartUpload call
+     */
+    public URIStatus checkIfComplete(String objectPath) {
+      try {
+        URIStatus objStatus = mUserFs.getStatus(new AlluxioURI(objectPath));
+        String uploadId = new String(objStatus.getXAttr()
+                .getOrDefault(PropertyKey.Name.S3_UPLOADS_ID_XATTR_KEY, new byte[0]));
+        if (objStatus.isCompleted() && StringUtils.equals(uploadId, mUploadId)) {
+          return objStatus;
+        }
+      } catch (IOException | AlluxioException ex) {
+        // can't validate if any previous attempt has succeeded
```

Comment:
maybe we can log the exception here in case we need more information sometimes

Reply:
will do

Reply:
done